stream_alert/alert_processor/main.py

```diff
@@ -46,350 +42,87 @@ def handler(event, context):
-                                               json.dumps(artifact),
-                                               headers,
-                                               False)
-                    if resp and resp.getcode() == 200:
-                        success_cnt += 1
-                logger.info('Successfully sent %s alerts to Phantom.', success_cnt)
+    # strip out unnecessary keys and sort
+    alert = sort_dict(alert)
 
-    def _s3(self, rule_name, alerts):
-        """Send alerts to an S3 bucket.
+    outputs = alert['metadata']['outputs']
+    # Get the output configuration for this rule and send the alert to each
+    for output in set(outputs):
+        try:
+            service, descriptor = output.split(':')
+        except ValueError as err:
+            LOGGER.error('outputs for rules must be declared with both a service and a '
+                         'descriptor for the integration (ie: \'slack:my_channel\')')
+            raise err
 
-        Organizes alerts into the following folder structure:
-            service/entity/rule_name/datetime.json
-        Each alert is a JSON object delimited by a newline.
-        """
-        alert_string = "\n".join([json.dumps(alert) for alert in alerts])
+        region = context.invoked_function_arn.split(':')[3]
+        function = context.invoked_function_arn.split(':')[-1]
 
-        client = boto3.client('s3', region_name=self.lambda_region)
-        resp = client.put_object(
-            Body=alert_string,
-            Bucket=self.bucket,
-            Key='{}/{}/{}/dt={}/stream_alerts_{}.json'.format(
-                # pull service/entity from the first alert.
-                # because logs are sent in groups, it's unlikely
-                # these values will be different across a group of alerts.
-                alerts[0]['metadata']['source']['service'],
-                alerts[0]['metadata']['source']['entity'],
-                rule_name,
-                datetime.now().strftime('%Y-%m-%d-%H-%M'),
-                datetime.now().isoformat('-')
-            )
-        )
-        logger.info('Alert sent to S3!')
+        # Retrieve the proper class to handle dispatching the alerts of this services
+        output_dispatcher = get_output_dispatcher(service, region, function)
 
-    # TODO(jacknagz): investigate message order bug
-    def _slack(self, rule_name, alerts):
-        """Send alert text to Slack
+        if not output_dispatcher:
+            continue
 
-        Args:
-            rule_name: The name of the triggered rule.
-            alerts: The array of alerts relevant to the triggered rule.
-        """
-        baseurl = self.creds.get('slack').url
-        token = self.creds.get('slack').secret
-        url = os.path.join(baseurl, token)
+        try:
+            output_dispatcher.dispatch(descriptor, rule_name, alert)
+        except BaseException as err:
+            LOGGER.error('an error occurred while sending alert to %s: %s',
```

Comment:
`LOGGER.exception('an error occurred while sending alert to %s', service)`
The exception context will be logged automatically